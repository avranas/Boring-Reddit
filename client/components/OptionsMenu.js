import React, { useEffect, useState } from "react";
import { useDispatch } from "react-redux";
import {
  getInitialThreadLimit,
  getInitialCommentsOn,
  getInitialHideNsfw,
} from "../slices/optionsSlice";

const OptionsMenu = (props) => {
  const [open, setOpen] = useState(false);
  const [threadLimit, setThreadLimit] = useState(getInitialThreadLimit());
  const [commentsOn, setCommentsOn] = useState(getInitialCommentsOn());
  const [hideNSFW, setHideNSFW] = useState(getInitialHideNsfw());

  const dispatch = useDispatch();
  const maxThreads = 25;

  const expand = () => {
    setOpen(true);
  };

  const close = (e) => {
    setOpen(false);
  };

  const handleChangeComments = (e) => {
    const commentsOn = e.target.checked;
    setCommentsOn(commentsOn);
    dispatch({ type: "options/setCommentsOn", payload: { commentsOn } });
  };

  const handleChangeNSFW = (e) => {
    const hideNSFW = e.target.checked;
    setHideNSFW(hideNSFW);
    dispatch({ type: "options/setHideNSFW", payload: { hideNSFW } });
  };

  const handleChangeThreadLimit = (e) => {
    let value = Number(e.target.value);
    if (value < 1) {
      value = 1;
    } else if (value > 25) {
      value = maxThreads;
    }
    setThreadLimit(value);
    dispatch({ type: "options/setThreadLimit", payload: { value } });
  };

  useEffect(() => {
    const handleClick = (e) => {
      const isDropdownButton = e.target.matches("[data-dropdown-button]");
      if (!isDropdownButton && e.target.closest("[data-dropdown") !== null)
        return;
      if (isDropdownButton) {
        if (open) {
          close();
        } else {
          expand();
        }
      } else {
        close();
      }
    };
    document.addEventListener("click", handleClick);
    return () => document.removeEventListener("click", handleClick);
  }, [open]);

  return (
    <div id="options" data-testid="options-menu">
      <button
        id="options-button"
        data-dropdown-button
        data-testid="options-menu-button"
      >
        Options
      </button>
      {open && (
        <div id="options-menu" data-dropdown>
          <div>
            <label htmlFor="thread-limit">Thread Limit:</label>
            <br />
            <input
              data-testid="thread-limit-input"
              type="number"
              id="thread-limit"
              name="thread-limit"
              min="1"
              max={maxThreads}
              defaultValue={threadLimit}
              onChange={handleChangeThreadLimit}
            />
          </div>
          <label className="form-check-label" htmlFor="flexSwitchCheckDefault">
            Comments On
          </label>
          <div className="form-check form-switch">
            <input
              data-testid="comment-switch"
              className="form-check-input"
              type="checkbox"
              role="switch"
              name="comments-on"
              id="comments-on"
              checked={commentsOn}
              onChange={handleChangeComments}
            />
          </div>
          <label className="form-check-label" htmlFor="flexSwitchCheckDefault">
            Hide NSFW
          </label>
          <div className="form-check form-switch">
            <input
              data-testid="nsfw-switch"
              className="form-check-input"
              type="checkbox"
              role="switch"
              name="hide-nsfw"
              id="hide-nsfw"
              checked={hideNSFW}
              onChange={handleChangeNSFW}
            />
          </div>
        </div>
      )}
    </div>
  );
};

export default OptionsMenu;
