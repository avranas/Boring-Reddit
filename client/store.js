import { configureStore } from "@reduxjs/toolkit";
import searchResultsReducer from "./slices/searchResultsSlice";
import currentPostReducer from "./slices/currentPostSlice";
import optionsReducer from "./slices/optionsSlice";

export default configureStore({
  reducer: {
    searchResults: searchResultsReducer,
    currentPost: currentPostReducer,
    options: optionsReducer,
  },
});
