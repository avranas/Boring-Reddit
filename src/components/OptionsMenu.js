import { useState } from "react";
import { useDispatch } from "react-redux";
import { getInitialThreadLimit, getInitialCommentsOff } from '../slices/optionsSlice';

const OptionsMenu = (props) => {
  
  const [ open, setOpen ] = useState(false);
  const [ threadLimit, setThreadLimit ] = useState(getInitialThreadLimit());
  const [ commentsOff, setCommentsOff ] = useState(getInitialCommentsOff());
  const dispatch = useDispatch();

  const toggleMenuOpen = () => {
    if (open) {
      setOpen(false);
    } else {
      setOpen(true);
    }
  }

  const handleSubmit = () => {
    setOpen(false);
    dispatch({ type: 'options/setThreadLimit', payload: { threadLimit }});
    dispatch({ type: 'options/setCommentsOff', payload: { commentsOff }});
  }

  const handleChangeComments = (e) => {
    setCommentsOff(e.target.checked);
  }

  const handleChangeThreadLimit = (e) => {
    let value = e.target.value
    if ( value < 1) {
      value = 1
    }
    setThreadLimit(Number(value));
  }



  return (
    <div>
      <div className="options-button">
        <button onClick={toggleMenuOpen}>
          Options
        </button>
      </div>
      {open && 
      <div className="options-menu">
        <label htmlFor="thread-limit">Thread Limit:</label><br/>
        <input type="number" id="thread-limit" name="thread-limit" min="1" max="25" defaultValue={threadLimit} onChange={handleChangeThreadLimit}/><br/>
        <label htmlFor="comments-off">Comments Off:</label><br/>
        <input type="checkbox" id="comments-off" name="comments-off" checked={commentsOff} onChange={handleChangeComments}/><br/>
        <button id="submit-options" onClick={handleSubmit}>Apply</button>
      </div>
      }
    </div>
  );
}

export default OptionsMenu;
