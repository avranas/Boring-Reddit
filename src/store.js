import { configureStore } from "@reduxjs/toolkit";
import searchResultsReducer from "../src/slices/searchResultsSlice";
import currentPostReducer from "../src/slices/currentPostSlice";
import optionsReducer from "../src/slices/optionsSlice";

export default configureStore({
  reducer: {
    searchResults: searchResultsReducer,
    currentPost: currentPostReducer,
    options: optionsReducer,
  },
});
