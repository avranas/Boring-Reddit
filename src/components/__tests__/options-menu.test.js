import { render, screen, cleanup } from '@testing-library/react';
import React from 'react';
import { BrowserRouter } from 'react-router-dom';
import OptionsMenu from '../OptionsMenu';
import { Provider } from 'react-redux';
import store from '../../store';
import renderer from 'react-test-renderer';

afterEach(() => {
  cleanup();
});

//const searchBar = <Provider store={store}><BrowserRouter><SearchBar/></BrowserRouter></Provider>;
const optionsMenu = <Provider store={store}><OptionsMenu/></Provider>;

test('should render OptionsMenu component', () => {
  render(optionsMenu);
  const optionsMenuElement = screen.getByTestId('options-menu');
  expect(optionsMenuElement).toBeInTheDocument();
})

test('OptionsMenu matches snapshot', () => {
  const tree = renderer.create(optionsMenu).toJSON();
  expect(tree).toMatchSnapshot();
});
