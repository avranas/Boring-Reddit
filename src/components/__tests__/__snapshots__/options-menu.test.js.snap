// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OptionsMenu matches snapshot 1`] = `
<div
  data-testid="options-menu"
  id="options"
>
  <button
    data-dropdown-button={true}
    id="options-button"
  >
    Options
  </button>
</div>
`;
